// basic-ftp/dist/FileInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FileType;
(function (FileType) {
    FileType[FileType["Unknown"] = 0] = "Unknown";
    FileType[FileType["File"] = 1] = "File";
    FileType[FileType["Directory"] = 2] = "Directory";
    FileType[FileType["SymbolicLink"] = 3] = "SymbolicLink";
})(FileType = exports.FileType || (exports.FileType = {}));
var FileInfo = /** @class */ (function () {
    function FileInfo(name) {
        this.name = "";
        this.type = FileType.Unknown;
        this.size = 0;
        this.permissions = { user: 0, group: 0, world: 0 };
        this.hardLinkCount = 0;
        this.link = "";
        this.group = "";
        this.user = "";
        this.date = "";
        this.name = name;
    }
    Object.defineProperty(FileInfo.prototype, "isDirectory", {
        get: function () {
            return this.type === FileType.Directory;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileInfo.prototype, "isSymbolicLink", {
        get: function () {
            return this.type === FileType.SymbolicLink;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FileInfo.prototype, "isFile", {
        get: function () {
            return this.type === FileType.File;
        },
        enumerable: true,
        configurable: true
    });
    FileInfo.Permission = {
        Read: 4,
        Write: 2,
        Execute: 1
    };
    return FileInfo;
}());
exports.FileInfo = FileInfo;
